CommandBlock API

main-- opened the pull request:
This adds a `BlockState` for command blocks, a command block change event and a command block execute event.

Ticket: https://bukkit.atlassian.net/browse/BUKKIT-3175
CraftBukkit PR: https://github.com/Bukkit/CraftBukkit/pull/957

Commit:
CommandBlock API

Change event, execute event and accessible as BlockState.

Wolvereness commented:
Format should never be 'null' and this needs a proper ticket. Also not sure if I agree with a command block execute event.

main-- commented:
@Wolvereness The block execute event would fix [BUKKIT-3165](https://bukkit.atlassian.net/browse/BUKKIT-3165). And why should format never be `null`?

Wolvereness commented:
You don't fixing a server command not firing an event by creating a 'new' event?

main-- commented:
Well, a command block command is basically not a server command. There is the `ServerCommandEvent` for console commands and the `PlayerCommandPreprocessEvent` for player commands, so adding a new event for block commands is reasonable.

riking commented on src/main/java/org/bukkit/event/block/CommandBlockChangeEvent.java:
```diff
@@ -0,0 +1,140 @@
+     */
+    public void setMessageFormat(String format) {
+        if (format != null)
+            String.format(format, command); // verify that they didn't mess up the format string
```
This can throw IllegalFormatException, you should mark as throwing, or something like that. Especially as you commented that it's to verify that they didn't mess it up. Also, may want to link to String.format in the javadoc, to make the formatting method clear.

I take it that setting to null is when you don't want a message to be sent, right? Why not just use the empty string for that?

main-- commented:
Changes (as @riking suggested):
- Improved javadoc for `setMessageFormat(String)`: Added `throws IllegalFormatException`, added a link to `String.format(String,Object...)`.
- "no message" is now represented by an empty string, a `null` format string is no longer allowed.

betterphp commented:
@Wolvereness This could have one important difference, the command block will execute a number of commands when it gets a signal due to expansion of the @ variables. In my opinion this event should provide the list of commands similar to the blockList() that you get from an EntityExplodeEvent. Yes that could be done with the redstone events but it would mean calling the command parsing a second time to get the list, adding some wasted time and a dependency on NMS.

riking commented:
If you still want this pulled, it needs to be updated to latest CraftBukkit.

riking commented:
We already have a CommandBlock blockstate now.

This is now defunct, please resubmit anything missing as a update to the current code.
